src/tr1/game/objects/traps/sliding_pillar.c

```diff
@@ -1,91 +1,252 @@
+    //     && M_IsAgainstCeiling(item)) {
+    //     const SECTOR *const sector = Room_GetWorldSector(
+    //         Room_Get(item->room_num), item->pos.x, item->pos.z);
+    //     if (item->pos.y < sector->floor.height) {
+    //         // If partially embedded from below e.g. jumping up into an
+    //         overhead
+    //         // block.
+    //         return height;
+    //     } else if (M_IsAgainstFloor(item)) {
+    //         // Clamped between floor and ceiling. Match up with similar case
+    //         in
+    //         // M_GetCeilingHeight to return same sentinel value;
+    //         return item->pos.y - WALL_L;
+    //     }
+    // }
+
+    // If under the bottom of the block.
+    if (y > item->pos.y) {
+        return height;
+    }
+
+    // If the the top of the block is under the floor height.
+    if (item->pos.y - WALL_L * 2 >= height) {
+        return height;
+    }
+
+    return item->pos.y - WALL_L * 2;
+}
+
+static int16_t M_GetCeilingHeight(
+    const ITEM *item, int32_t x, int32_t y, int32_t z, int16_t height)
+{
+    if (item->status == IS_INVISIBLE) {
+        return height;
+    }
+
+    if (item->current_anim_state == PILLAR_STATE_MOVING) {
+        return height;
+    }
+
+    // Only care if we are inside the block footprint.
+    if (!M_IsItemOnTop(item, x, z)) {
+        return height;
+    }
+
+    // if (y <= item->pos.y && y > item->pos.y - WALL_L && !item->gravity) {
+    //     if (M_IsAgainstCeiling(item)) {
+    //         // If clamped betwee floor and ceiling return same sentinel value
+    //         as
+    //         // M_GetFloorHeight.
+    //         return M_IsAgainstFloor(item) ? item->pos.y - WALL_L :
+    //         item->pos.y;
+    //     }
+    //     return height;
+    // }
+
+    // If above the top of the block.
+    if (y <= item->pos.y - WALL_L * 2) {
+        return height;
+    }
+
+    // If the the bottom of the block is above the ceiling height.
+    if (item->pos.y <= height) {
+        return height;
+    }
+
+    return item->pos.y;
+}
+
 static void M_Setup(OBJECT *const obj)
 {
     obj->initialise_func = M_Initialise;
-    obj->handle_flip_func = M_HandleFlip;
     obj->handle_save_func = M_HandleSave;
     obj->control_func = M_Control;
+    obj->floor_height_func = M_GetFloorHeight;
+    obj->ceiling_height_func = M_GetCeilingHeight;
     obj->save_position = true;
     obj->save_anim = true;
     obj->save_flags = true;
+    obj->add_walkable_func = M_AddWalkable;
 }
 
 static void M_Initialise(const int16_t item_num)
 {
     ITEM *const item = Item_Get(item_num);
-    Room_AlterFloorHeight(item, -WALL_L * 2);
-}
-
-static void M_HandleFlip(ITEM *const item, const ROOM_FLIP_STATUS flip_status)
-{
-    if (flip_status == RFS_FLIPPED) {
-        Room_AlterFloorHeight(item, -WALL_L * 2);
-    } else {
-        Room_AlterFloorHeight(item, WALL_L * 2);
-    }
+    // Room_AlterFloorHeight(item, -WALL_L * 2);
```

Comment:
The sliding pillar will need a `MovableBlock_UpdateBox` call in its initialise function to match OG. I've updated the test level with blockable sectors under 3 of the 4 sliding pillars.

As an aside, perhaps in the future after further merging `MovableBlock_UpdateBox` could be moved somewhere else common.